pkg/operator/configobserver/apiserver/observe_audit.go

```diff
@@ -0,0 +1,88 @@
+package apiserver
+
+import (
+	"fmt"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog"
+
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+// AuditPolicyPathGetter allows the observer to be agnostic of the source of audit profile(s).
+type AuditPolicyPathGetter interface {
+	// AuditPolicyPathGetter returns the path to the audit policy file (associated with the
+	// given profile) in the static manifest folder.
+	GetAuditPolicyPath(profile string) (string, error)
+
+	// GetDefaultAuditPolicyPath returns the path to the default audit policy.
+	GetDefaultAuditPolicyPath() string
+}
+
+// NewAuditObserver returns an ObserveConfigFunc that observes the audit field of the APIServer resource
+// and sets the apiServerArguments:audit-policy-file field for the apiserver appropriately.
+func NewAuditObserver(pathGetter AuditPolicyPathGetter) configobserver.ObserveConfigFunc {
+	var (
+		apiServerArgumentsAuditPath = []string{"apiServerArguments", "audit-policy-file"}
+	)
+
+	return func(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+		errs := []error{}
+
+		// does the current config already have an audit policy path set? if so we should make an effort to
+		// return it as the observed config in case of error, otherwise we resort to the default policy path.
+		auditPolicyPath := pathGetter.GetDefaultAuditPolicyPath()
+		current, _, err := unstructured.NestedStringSlice(existingConfig, apiServerArgumentsAuditPath...)
+		if err != nil {
+			// best-effort approach, let's keep going.
+			errs = append(errs, err)
+		}
+		if len(current) > 0 {
+			auditPolicyPath = current[0]
+		}
+
+		defaultObservedConfig := map[string]interface{}{}
+		if err := unstructured.SetNestedStringSlice(defaultObservedConfig, []string{auditPolicyPath}, apiServerArgumentsAuditPath...); err != nil {
+			// If we are here, then we are not setting any default for observed audit profile.
+			// looks like returning the existing config is the best alternative here. This puts a constraint on the
+			// apiserver to have a default configuration in its bindata.
+			return existingConfig, append(errs, err)
+		}
+
+		listers := genericListers.(APIServerLister)
+		apiServer, err := listers.APIServerLister().Get("cluster")
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				klog.Warningf("apiserver.config.openshift.io/cluster: not found")
+
+				// return the audit policy path from the current config if set, otherwise we use the default configuration.
+				return defaultObservedConfig, errs
+			}
+
+			return defaultObservedConfig, append(errs, err)
```

Comment:
on non-not-found errors we return the existing config. This here would change the config for the apiserver. We don't want that. It creates flapping on error situations.

Reply:
on any lister error, the observer function is returning the current/existing config.